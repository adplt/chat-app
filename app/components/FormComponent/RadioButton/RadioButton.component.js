import React, {PureComponent} from 'react';
import {View, Text, WebView} from 'react-native';
import {RadioButtons} from 'react-native-radio-buttons';
import styles from './RadioButton.styles';
import Touchable from '../../TouchableComponent/Touchable.component';
import noop from 'lodash/noop';
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';

export default class RadioButton extends PureComponent {

  static propTypes = {
    options: PropTypes.array,
    input: PropTypes.object,
    onPressRadioButton: PropTypes.func,
    warningText: PropTypes.string,
    renderContainer: PropTypes.object,
    renderRow: PropTypes.bool,
    label: PropTypes.string,
    labelStyle: PropTypes.object,
    value: PropTypes.any,
  }

  static defaultProps = {
    options: [], // option = [{label: '', value: ''}, ...]
    input: {
      onChange: noop,
    }
  }

  setSelectedOption = (selectedOption, option) => option.value === this.props.input.value.value;

  renderOption = (option, selected, onSelect, index) => {
    const {onPressRadioButton = noop, warningText = ''} = this.props;
    return (<Touchable onPress={onSelect} key={index}>
      <View>
        <View style={styles.optionStyle}>
          <View style={styles.buttonStyle}>{selected && <View style={styles.buttonActive} />}</View>
          <View style={styles.labelContainer}>
            <Text style={styles.textStyle} onPress={onPressRadioButton}>{option.label}</Text>
            <Text style={styles.textStyle}>{option.sublabel}</Text>
          </View>
        </View>
        {
          warningText !== '' &&
          <View style={styles.webViewContainer}>
            <WebView style={styles.webViewStyle} source={{html: warningText}}/>
          </View>
        }
      </View>
    </Touchable>);
  }

  renderOptionRow = (option, selected, onSelect, index) => {
    const {onPressRadioButton = noop, warningText = ''} = this.props;
    return (
      <View style={styles.halfWidth} key={index}>
        <Touchable onPress={onSelect}>
          <View>
            <View style={styles.optionStyle}>
              <View style={styles.buttonStyle}>{selected && <View style={styles.buttonActive} />}</View>
              <View style={styles.labelContainer}>
                <Text style={styles.textStyle} onPress={onPressRadioButton}>{option.label}</Text>
                <Text style={styles.textStyle}>{option.sublabel}</Text>
              </View>
            </View>
            {
              warningText !== '' &&
              <View style={styles.webViewContainer}>
                <WebView style={styles.webViewStyle} source={{html: warningText}}/>
              </View>
            }
          </View>
        </Touchable>
      </View>
    );
  }

  renderContainer = (optionNodes) => <View>{optionNodes}</View>

  renderContainerRow = (optionNodes) => <View style={styles.renderContainerRow}>{optionNodes}</View>

  render () {
    const {options, input, renderRow = false, label, labelStyle, value} = this.props;
    return (
      <View>
        {
          !isEmpty(label) &&
            <Text style={[styles.label, labelStyle]}>{label}</Text>
        }
        <View style={styles.radioButton}>
          <RadioButtons
            options={options}
            onSelection={input.onChange}
            selectedOption={value}
            renderOption={renderRow ? this.renderOptionRow : this.renderOption}
            renderContainer={renderRow ? this.renderContainerRow : this.renderContainer}
            testOptionEqual={this.setSelectedOption}
          />
        </View>
      </View>
    );
  }
}
