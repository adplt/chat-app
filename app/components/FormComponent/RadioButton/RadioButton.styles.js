import {theme} from '../../../styles/core.style';
import {Dimensions} from 'react-native';
import {fontSizeLarge, fontWeightBold} from '../../../styles/common.style';

const {width} = Dimensions.get('window');

export default {
  optionStyle: {
    flexDirection: 'row',
    paddingTop: 10
  },
  buttonStyle: {
    height: 20,
    width: 20,
    borderRadius: 12,
    borderWidth: 2,
    borderColor: theme.brand,
    alignItems: 'center',
    justifyContent: 'center'
  },
  buttonActive: {
    height: 12,
    width: 12,
    borderRadius: 6,
    backgroundColor: theme.brand
  },
  labelContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  textStyle: {
    fontSize: theme.fontSizeNormal,
    paddingLeft: 5
  },
  subtextStyle: {
    fontSize: theme.fontSizeSmall,
    paddingRight: 5
  },
  webViewContainer: {
    paddingLeft: 20,
  },
  webViewStyle: {
    height: 90,
  },
  halfWidth: {
    width: (width - 40) / 2
  },
  renderContainerRow: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center'
  },
  label: {
    ...fontSizeLarge,
    ...fontWeightBold,
    marginTop: 10,
    marginBottom: 5,
  },
  radioButton: {
    marginLeft: 20,
  },
};
