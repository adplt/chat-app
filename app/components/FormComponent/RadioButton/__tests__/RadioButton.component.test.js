import React from 'react';
import renderer from 'react-test-renderer';
import RadioButton from '../RadioButton.component';
import {shallow, configure} from 'enzyme';
import noop from 'lodash/noop';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('Radio Button component', () => {
  it('renders correctly', () => {
    const options = [{label: 'xxx', value: 'xxx'}];
    const input = {value: {value: 'xxx'}, onChange: noop};
    const tree = renderer.create(<RadioButton options={options} input ={input}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('it should return true for the selected option', () => {
    const input = {value: {value: 'xxx'}};
    const selectedOption = {value: 'xxx'};
    const wrapper = shallow(<RadioButton input={input}/>);
    expect(wrapper.instance().setSelectedOption(selectedOption, selectedOption)).toBeTruthy();
  });
});
