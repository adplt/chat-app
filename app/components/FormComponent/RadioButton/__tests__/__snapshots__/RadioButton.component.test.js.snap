// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Button component renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "marginLeft": 20,
      }
    }
  >
    <View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "paddingTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderColor": "#392d2c",
                  "borderRadius": 12,
                  "borderWidth": 2,
                  "height": 20,
                  "justifyContent": "center",
                  "width": 20,
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "#392d2c",
                    "borderRadius": 6,
                    "height": 12,
                    "width": 12,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                onPress={[Function]}
                style={
                  Object {
                    "fontSize": 17,
                    "paddingLeft": 5,
                  }
                }
              >
                xxx
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "fontSize": 17,
                    "paddingLeft": 5,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
